packages/back-end/src/events/handlers/utils.ts

```diff
@@ -0,0 +1,182 @@
+import uniq from "lodash/uniq";
+import isEqual from "lodash/isEqual";
+import intersection from "lodash/intersection";
+import {
+  FeatureUpdatedNotificationEvent,
+  NotificationEvent,
+} from "../notification-events";
+import { ApiFeature } from "../../../types/openapi";
+
+export type FilterDataForNotificationEvent = {
+  tags: string[];
+  projects: string[];
+};
+
+export const getFilterDataForNotificationEvent = (
+  event: NotificationEvent
+): FilterDataForNotificationEvent | null => {
+  let invalidEvent: never;
+
+  switch (event.event) {
+    case "user.login":
+      return null;
+
+    case "feature.created":
+      return {
+        tags: event.data.current.tags || [],
+        projects: event.data.current.project
+          ? [event.data.current.project]
+          : [],
+      };
+
+    case "feature.updated":
+      return {
+        tags: uniq(
+          (event.data.current.tags || []).concat(event.data.previous.tags || [])
+        ),
+        projects: uniq(
+          (event.data.current.project
+            ? [event.data.current.project]
+            : []
+          ).concat(
+            event.data.previous.project ? [event.data.previous.project] : []
+          )
+        ),
+      };
+
+    case "feature.deleted":
+      return {
+        tags: event.data.previous.tags || [],
+        projects: event.data.previous.project
+          ? [event.data.previous.project]
+          : [],
+      };
+
+    case "experiment.created":
+      return {
+        tags: event.data.current.tags || [],
+        projects: event.data.current.project
+          ? [event.data.current.project]
+          : [],
+      };
+
+    case "experiment.updated":
+      return {
+        tags: uniq(
+          (event.data.current.tags || []).concat(event.data.previous.tags || [])
+        ),
+        projects: uniq(
+          (event.data.current.project
+            ? [event.data.current.project]
+            : []
+          ).concat(
+            event.data.previous.project ? [event.data.previous.project] : []
+          )
+        ),
+      };
+
+    case "experiment.deleted":
+      return {
+        tags: event.data.previous.tags || [],
+        projects: event.data.previous.project
+          ? [event.data.previous.project]
+          : [],
+      };
+
+    default:
+      invalidEvent = event;
+      throw `Invalid event: ${invalidEvent}`;
+  }
+};
+
+export const filterEventForEnvironments = ({
+  event,
+  environments,
+}: {
+  event: NotificationEvent;
+  environments: string[];
+}): boolean => {
+  let invalidEvent: never;
+
+  switch (event.event) {
+    case "user.login":
+      return false;
+
+    case "experiment.created":
+    case "experiment.updated":
+    case "experiment.deleted":
+      return true;
+
+    case "feature.created":
+    case "feature.deleted":
+      return true;
+
+    case "feature.updated":
+      return filterFeatureUpdatedNotificationEventForEnvironments({
```

Comment:
might be good to leave a comment here, since it kinda threw me off why we usually just return "true"

```
// Will only notify for environments in which rules were modified. Other feature events (currently) apply to all environments
```